### PR TITLE
Fixed #30108 -- Allowed adding foreign key constraints in the same statement that adds a field.

### PR DESCRIPTION
**Pull request by dtao**
Curiously, the `BaseDatabaseSchemaEditor` class already supported the attribute `sql_create_inline_fk`, but the only implementation that supported it was sqlite3. Moreover, this attribute was only used to create FK constraints inline during table creation, not when adding a new field to an existing table.

This introduces another attribute, `sql_add_field_inline_fk`, for creating FK constraints inline when adding fields. The sqlite3 engine supports this, and so does PostgreSQL.

It was not trivially easy to enable this functionality for table creation in PostgreSQL, as explained in a [comment on #30108](https://code.djangoproject.com/ticket/30108#comment:1):

> Unfortunately this ended up causing test failures because there are ​test models defined with a cyclic dependency. It seems that SQLite doesn't mind if you create a table with a foreign key to another table that doesn't exist yet, but PostgreSQL does.

Fortunately, creating FK constraints inline on table _creation_ is hardly critical, as the default behavior of doing so in a deferred query, thereby triggering a table scan, is effectively benign for an empty table.

### CODE REVIEW

**Comment by charettes** [comment]
> ... but the only implementation that supported it was sqlite

The reason for this is that SQLite simply does't support `ALTER TABLE ADD CONSTRAINT` (actually it doesn't support most form of `ALTER TABLE`) and requires such constraints to be inlined.

> It was not trivially easy to enable this functionality for table creation in PostgreSQL

yeah circular foreign keys are tricky. Ideally the auto-detector would never generate circular `CreateModel` with foreign keys and make sure to create one of the model with a base field and add a single `AlterField('foo', ForeignKey)` after so `SchemEditor.create_model` doesn't have to rely on `deferred_sql`. I'm not even sure it's smart enough to do that now when a new circular relationship is defined in the same app. Circular relationships across apps are event trickier as they require at least three migrations to be created to break the cycle (a first with a `CreateModel('Foo', bar=IntegerField)`, a second with `CreateModel('Bar', ForeignKey('bar'))`, and a last one with `AlterField('Foo', bar=ForeignKey('foo')) with appropriate `dependencies`. Lastly even even if we were to fix the auto-detector to correctly resolve these cycles into the appropriate migrations turning the feature on would break historical migrations.

In summary, turning this feature on for `AddField` is safe but as you've come to discover it's trickier for `CreateModel` operations.

File: django/db/backends/sqlite3/schema.py [hunk 1]
```diff
@@ -14,6 +14,7 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     sql_delete_table = "DROP TABLE %(table)s"
     sql_create_fk = None
     sql_create_inline_fk = "REFERENCES %(to_table)s (%(to_column)s) DEFERRABLE INITIALLY DEFERRED"
+    sql_add_field_inline_fk = "REFERENCES %(to_table)s (%(to_column)s) DEFERRABLE INITIALLY DEFERRED"
```
Comment (charettes) [inline django/db/backends/sqlite3/schema.py]: This is unnecessary as add field operations result in table rebuilds and `sql_create_inline_fk` usage. See `add_field` below.

**Comment by dtao** [comment]
> Any reason why this can't be turned on for MySQL and Oracle as well?

Not that I'm aware of. I am less familiar with those database engines, so I didn't touch them. I'm happy to add the new attribute to their respective schema editor classes if that's something I should do.

**Comment by charettes** [comment]
> I'm happy to add the new attribute to their respective schema editor classes if that's something I should do.

I think it might be problematic for MySQL as it doesn't support deferreable constraints but it's worth a shot.

**Comment by charettes** [comment]
buildbot, test on oracle.

**Comment by dtao** [comment]
Well, I see that the Oracle build failed on [two](https://djangoci.com/job/pull-requests-oracle/database=oracle18,label=bionic-pr,python=python3.7/287/testReport/junit/migrations.test_operations/OperationTests/test_remove_fk/) [test cases](https://djangoci.com/job/pull-requests-oracle/database=oracle18,label=bionic-pr,python=python3.7/287/testReport/junit/model_options.test_tablespaces/TablespacesTests/test_tablespace_for_many_to_many_field/) with the same error.

    ORA-12847: retry parsing due to concurrent DDL operation

Unfortunately I don't know what this means :( Google surprisingly offers very little insight. Is it possible there is something flaky about these tests (i.e., they might pass if run again), or does this indicate that there's definitely something wrong that I should fix?

**Comment by charettes** [comment]
I assume this Oracle error is similar to the _pending trigger_ error that happens on PostgreSQL when you add a deferrable foreign key without performing a check constraint. Lets skip it for now.

Let's remove the Oracle support attempt code and and squash all commits under a single one with the PR's title as message. This should eventually get merged by fellows, thank you for your contribution.

**Comment by charettes** [comment]
Please uncheck _patch needs improvement_ once all comments have been addressed.

**Comment by felixxm** [comment]
@charettes @dtao Oracle failures are not related with this PR, they appeared also in the [master build](https://djangoci.com/job/django-oracle/lastCompletedBuild/database=oracle18,python=python3.7/testReport/migrations.test_operations/OperationTests/test_rename_m2m_model_after_rename_field/). I will try to fix them later.

File: django/db/backends/base/schema.py [hunk 2]
```diff
@@ -77,6 +77,7 @@ class BaseDatabaseSchemaEditor:
         "REFERENCES %(to_table)s (%(to_column)s)%(deferrable)s"
     )
     sql_create_inline_fk = None
+    sql_add_field_inline_fk = None
```
[the remark this answers is not in the document]
Comment (dtao) [reply django/db/backends/base/schema.py]: Good call—I hadn't noticed that. I think I'll go with `sql_create_column_inline_fk` (since `add` is _also_ not consistent with any of the other attribute names).

File: tests/indexes/tests.py [hunk 3]
```diff
@@ -226,11 +226,9 @@ def test_no_index_for_foreignkey(self):
...
                 editor.add_field(ArticleTranslation, new_field)
                 field_created = True
-                self.assertEqual([str(statement) for statement in editor.deferred_sql], [
-                    'ALTER TABLE `indexes_articletranslation` '
-                    'ADD CONSTRAINT `indexes_articletrans_new_foreign_key_id_d27a9146_fk_indexes_a` '
-                    'FOREIGN KEY (`new_foreign_key_id`) REFERENCES `indexes_article` (`id`)'
-                ])
+                # There should be no deferred SQL as the FK constraint was
+                # included inline in the statement to add the field.
+                self.assertEqual([str(statement) for statement in editor.deferred_sql], [])
```
[the remark this answers is not in the document]
Comment (dtao) [reply tests/indexes/tests.py]: Reasonable.

File: django/db/backends/base/schema.py [hunk 4]
```diff
@@ -433,6 +434,21 @@ def add_field(self, model, field):
...
+                to_column = field.remote_field.model._meta.get_field(field.remote_field.field_name).column
+                definition += " " + self.sql_add_field_inline_fk % {
+                    "name": self._fk_constraint_name(model, field, constraint_suffix),
+                    "column": self.quote_name(field.column),
+                    "to_table": self.quote_name(to_table),
+                    "to_column": self.quote_name(to_column),
+                }
+            # Otherwise, add FK constraints later
+            else:
+                self.deferred_sql.append(self._create_fk_sql(model, field, constraint_suffix))
```
[the remark this answers is not in the document]
Comment (dtao) [reply django/db/backends/base/schema.py]: I don't _think_ it should matter? My understanding of how this works is that this is simply appending to a list of SQL statements which get executed [when the context manager exits](https://github.com/django/django/blob/master/django/db/backends/base/schema.py#L109-L110)—and only then if no exception were raised.

It *could* have made a difference if it weren't for the exception check, as previously an exception being raised in `execute` would prevent the deferred SQL statement from even being appended to the list. But given that [it won't be executed anyway if there's an exception](https://github.com/django/django/blob/master/django/db/backends/base/schema.py#L108), I believe there is functionally no difference here.

Anyway... all of the tests pass ;)

All that said, if you'd prefer, I could finesse this code a bit so that the default cause (where `sql_create_column_inline_fk` isn't defined) is truly identical to the way it was before. I personally think it would make the code a bit more complicated than it needs to be, but it isn't exactly the hill I'd choose to die on.

**Comment by dtao** [comment]
I also see that there are a couple of failing build, but they _appear_ to be unrelated to these changes—perhaps the result of ephemeral issues?

File: django/db/backends/base/schema.py [hunk 5]
```diff
@@ -433,6 +434,21 @@ def add_field(self, model, field):
         db_params = field.db_parameters(connection=self.connection)
         if db_params['check']:
             definition += " " + self.sql_check_constraint % db_params
+        if field.remote_field and self.connection.features.supports_foreign_keys and field.db_constraint:
+            constraint_suffix = "_fk_%(to_table)s_%(to_column)s"
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: Please use single quotes.

File: django/db/backends/base/schema.py [hunk 6]
```diff
@@ -433,6 +434,21 @@ def add_field(self, model, field):
...
+            # Add FK constraint inline, if possible
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: Trailing dot is missing.

File: django/db/backends/base/schema.py [hunk 7]
```diff
@@ -433,6 +434,21 @@ def add_field(self, model, field):
...
+            if self.sql_create_column_inline_fk:
+                to_table = field.remote_field.model._meta.db_table
+                to_column = field.remote_field.model._meta.get_field(field.remote_field.field_name).column
+                definition += " " + self.sql_create_column_inline_fk % {
+                    "name": self._fk_constraint_name(model, field, constraint_suffix),
+                    "column": self.quote_name(field.column),
+                    "to_table": self.quote_name(to_table),
+                    "to_column": self.quote_name(to_column),
+                }
+            # Otherwise, add FK constraints later
```
Comment (felixxm) [inline django/db/backends/base/schema.py]: Trailing dot is missing.

File: django/db/backends/postgresql/schema.py [hunk 8]
```diff
@@ -15,6 +15,8 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     sql_create_index = "CREATE INDEX %(name)s ON %(table)s%(using)s (%(columns)s)%(extra)s%(condition)s"
     sql_delete_index = "DROP INDEX IF EXISTS %(name)s"
 
+    sql_create_column_inline_fk = "REFERENCES %(to_table)s (%(to_column)s) DEFERRABLE INITIALLY DEFERRED"
```
[the remark this answers is not in the document]
Comment (dtao) [reply django/db/backends/postgresql/schema.py]: Ah, so _that's_ how that works ;)

I think I understand. I'll make the change.

File: django/db/backends/oracle/schema.py [hunk 9]
```diff
@@ -15,6 +15,7 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
     sql_alter_column_default = "MODIFY %(column)s DEFAULT %(default)s"
     sql_alter_column_no_default = "MODIFY %(column)s DEFAULT NULL"
     sql_delete_column = "ALTER TABLE %(table)s DROP COLUMN %(column)s"
+    sql_create_column_inline_fk = "CONSTRAINT %(name)s REFERENCES %(to_table)s(%(to_column)s)"
```
[the remark this answers is not in the document]
Comment (dtao) [reply django/db/backends/oracle/schema.py]: That sounds right to me too. I will add it.

File: django/db/backends/postgresql/schema.py [hunk 10]
```diff
@@ -15,6 +15,8 @@ class DatabaseSchemaEditor(BaseDatabaseSchemaEditor):
...
+
```
[the remark this answers is not in the document]
Comment (dtao) [reply django/db/backends/postgresql/schema.py]: OK. I guess the code paragraphs here are organized according to the type of thing being created/deleted? This one's a gray area for me because, although the attribute name ends in `_fk` like the attribute below, this particular attribute relates to creating a FK constraint _while_ creating a column.

Not that it really matters. I'll delete the line.

File: django/db/backends/base/schema.py (hunk 5 above)
Comment (dtao) [reply django/db/backends/base/schema.py]: OK. For what it's worth, this file is all over the place in that respect... in fact it seems _most_ of the strings use double quotes (I was doing my best to follow convention, I swear). But I can obviously change this to use single quotes.

File: django/db/backends/base/schema.py (hunk 6 above)
Comment (dtao) [reply django/db/backends/base/schema.py]: If you insist. Again, I was doing my best to be consistent with the surrounding code. None of the other comments for this particular function end with a period.

File: django/db/backends/base/schema.py (hunk 7 above)
Comment (dtao) [reply django/db/backends/base/schema.py]: See my previous comment. Would you like me to:

1. Add a trailing dot in the two places you indicated, so that _some_ of the comments in this function end in a period and others don't?
2. Update _all_ of the comments in this function, making changes to lines of code (and arguably polluting the commit history) that aren't related to this pull request?
3. Something else?

File: tests/indexes/tests.py [hunk 11]
```diff
@@ -226,11 +226,9 @@ def test_no_index_for_foreignkey(self):
                 new_field.set_attributes_from_name('new_foreign_key')
                 editor.add_field(ArticleTranslation, new_field)
                 field_created = True
-                self.assertEqual([str(statement) for statement in editor.deferred_sql], [
-                    'ALTER TABLE `indexes_articletranslation` '
-                    'ADD CONSTRAINT `indexes_articletrans_new_foreign_key_id_d27a9146_fk_indexes_a` '
-                    'FOREIGN KEY (`new_foreign_key_id`) REFERENCES `indexes_article` (`id`)'
-                ])
+                # There should be no deferred SQL as the FK constraint was
```
Comment (timgraham) [inline tests/indexes/tests.py]: It looks to me like this test should verify the SQL rather than verifying that there isn't any deferred SQL.

**Comment by charettes** [comment]
> Are you happy without any tests? It seems to me a sql_create_column_inline_fk attribute could be inadvertently removed and cause this to silently revert to the "add FK constraints later" approach.

Right, I guess some schema add field tests could be adjusted to make sure no deferred SQL is added.

File: django/db/backends/base/schema.py (hunk 7 above)
Comment (felixxm) [reply django/db/backends/base/schema.py]: Thanks for your effort :+1:. It is a kindly request for the first option. Trailing dot should be at the end of a sentence but there is no point in doing unrelated refactoring of existing code.

File: django/db/backends/base/schema.py (hunk 5 above)
Comment (felixxm) [reply django/db/backends/base/schema.py]: We prefer single quotes, but again there is no point in doing unrelated refactoring of existing code (see [coding-style](
https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style)). Just a nitpick.

Comment (dtao) [reply django/db/backends/base/schema.py]: This is done.

File: django/db/backends/base/schema.py (hunk 6 above)
Comment (dtao) [reply django/db/backends/base/schema.py]: Fixed.

File: django/db/backends/base/schema.py (hunk 7 above)
Comment (dtao) [reply django/db/backends/base/schema.py]: Fixed.

**Comment by dtao** [comment]
At the risk of derailing this pull request, I've just pushed another commit introducing a test for this new behavior. I have verified that the test fails (with a descriptive error message) prior to these changes.

This project has lots of tests! It was a little intimidating trying to figure out where this test should go. I settled on tests/migrations/test_operations.py, where there were already tests for adding various types of fields with a moderate amount of existing code that looked reusable to me.

I hope this does not open up a whole new can of worms. I do think @timgraham [raised a good point](https://github.com/django/django/pull/10874#pullrequestreview-194784611) and that a test is warranted to ensure this behavior doesn't accidentally get reverted down the road.

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by dtao** [comment]
I see all the tests are passing, including Oracle. Should I squash this down to 1 commit again?

File: tests/migrations/test_operations.py [hunk 12]
```diff
@@ -1094,6 +1094,41 @@ def test_add_binaryfield(self):
         self.assertEqual(bytes(pony.digits), b"42")
         self.assertEqual(bytes(pony.quotes), b'"\'"')
 
+    def test_add_foreignkey(self):
+        """
+        Tests that adding a ForeignKey creates the corresponding constraint.
```
Comment (timgraham) [inline tests/migrations/test_operations.py]: Docstrings should state the expected behavior and omit prefixes like "Tests that" since all tests test things. Ticket references should be reserved for obscure issues (not needed here). These guidelines are from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: tests/migrations/test_operations.py [hunk 13]
```diff
@@ -1094,6 +1094,41 @@ def test_add_binaryfield(self):
...
+                'stable',
+                models.ForeignKey('test_adfkfl.Stable', on_delete=models.CASCADE),
+            )
+        ]
+        with connection.schema_editor() as editor:
+            migration.apply(project_state, editor)
+
+        self.assertFKExists('test_adfkfl_pony', ['stable_id'], ('test_adfkfl_stable', 'id'))
+
+        if connection.vendor in ('mysql', 'oracle', 'postgresql'):
```
Comment (timgraham) [inline tests/migrations/test_operations.py]: This should use a database feature `django/db/backends/base/features.py` so that other third-party database backends can also use the test.

Comment (dtao) [reply tests/migrations/test_operations.py]: I wondered if someone was going to say that!

Any opinion what the feature should be named? Maybe `can_create_fk_constraints_inline`?

File: tests/migrations/test_operations.py [hunk 14]
```diff
@@ -1094,6 +1094,41 @@ def test_add_binaryfield(self):
...
+            deferred_sql = [str(statement) for statement in editor.deferred_sql]
+            deferred_add_constraint = [
+                sql for sql in deferred_sql if sql.startswith('ALTER TABLE') and 'ADD CONSTRAINT' in sql
+            ]
+            self.assertFalse(
+                deferred_add_constraint,
+                'The %s backend supports creating FK constraints inline when '
+                'adding columns; deferred SQL to add the constraint is '
+                'unnecessary (#30108).' % connection.vendor
```
Comment (charettes) [inline tests/migrations/test_operations.py]: No need to refer to the ticket here.

File: tests/migrations/test_operations.py (hunk 13 above)
Comment (timgraham) [reply tests/migrations/test_operations.py]: `can_create_inline_fk` would mirror the name of `sql_create_inline_fk`.

**Comment by dtao** [comment]
@timgraham said:

> If it's possible to put the test in tests/schema, that may be a better location

This is done. Turns out it wasn't too hard.

**Comment by dtao** [comment]
I'm not sure what the test failure's about. As far as I can tell, it's unrelated.

```
PermissionError: [Errno 13] Permission denied: 'C:\\Jenkins\\workspace\\pull-requests-windows\\database\\sqlite3\\label\\windows\\python\\Python37\\tests\\template_tests\\templates\\first'
```

**Comment by dtao** [comment]
\o/
